### DISCUSSION TITLE
Why do we need the "Eclipse-Testing-Framework"

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-platform/eclipse.platform.releng#149] [Ideas]
In the past there was always some issues with the "Eclipse-Testing-Framework"

- https://wiki.eclipse.org/Platform-releng/Eclipse_Test_Framework
- https://wiki.eclipse.org/Platform-releng/Automated_Testing

Some of the issue include:

- Tests are not executed because they are not added to a special "AllTests"
- Dependencies are not properly fetched
- People are confused because we have a different (ant?) setup than in the verification builds
- ...

The recent discussions about JDT development should become more "modern" has bring this back into my mind that I always have wondered **why** "Eclipse-Testing-Framework" is required?

1. We are able to execute the test in regular maven verification builds (GitHub and Jenkins)
2. We can execute Eclipse Installations with Eclipse Run Mojo
3. We can even consume additional artifacts in Tycho builds

So what is the blocker here? Do we need any support from Tycho/Maven to maybe execute tests without compile them? Is running them "inside" the SDK zip an issue? Maybe something else?

As a very native aproach I would say we just need a Mojo that runs an Eclipse (isn't uiharness the way to do so already?) and probably select some (or all?) test-fragments and execute the precompiled ones with the current running system.

I'm specifically asking because I plan to improve testing-support for Tycho and covering this use-case would be a good addition here, but as mentioned I'm not really sure if there is actually something _missing_ or if it is just no one wants to put _effort_ in migration this stuff?

**Comment by akurtakov** [comment]
Being able to run test bundles' test suites on current system with predefined product on different os/java combos each bundle in clean osgi environment while skipping the compilation and packaging parts which Maven/Tycho kind of enforce now. As the tests are already terribly slow (up to 9 h on some combos) it's critical to not do the compile/package steps as that would seriously question the ability to have daily runs.
The rest are things like defining which test suite and etc. to run but this is already available on Maven side. An extra entry point/profile/whatever might be needed to define what to run but here we go into dirty details.
I'm really looking forward to retire this part of our code base but unfortunately it's last in the todo list.

**Reply by laeubi** [reply]
> Being able to run test bundles' test suites on current system with predefined product on different os/java combos each bundle in clean osgi environment while skipping the compilation and packaging parts which Maven/Tycho kind of enforce now.

That's a part I really not sure about (and why I'm asking here), at least from the code perspective I don't se a reason why compile (or package) should be required at all.

So (possible to simple) assumption here is, that the only "issue" is that test-plugin is scanning the modules target dir, but instead should scan the classpath (or better the dependency path) and we should all be set? Maybe define used SDK zip anywhere? 

> The rest are things like defining which test suite and etc. to run but this is already available on Maven side. 

Thats something I'm think will follow the BND (and PAX Exam is using that as well) to just record the tests in the package phase (e.g. as a file or header) that later is then investigated. But even I think this can be done dynamically if one knows what are the jars to scan (se idea above with having a dependency section) ...

> An extra entry point/profile/whatever might be needed to define what to run but here we go into dirty details. I'm really looking forward to retire this part of our code base but unfortunately it's last in the todo list.

My first goal would be to replace *one* test fragment with the new approach that seems doable and then it is mostly a matter of transforming the others, but I first need to understand the quirks that might interfere here. So if you are aware of a very simply example let me know, I though Equinox might be good but I havent checked if equinox tests are actually executed that way...

**Comment by mickaelistria** [comment]
> That's a part I really not sure about (and why I'm asking here), at least from the code perspective I don't se a reason why compile (or package) should be required at all.

Compiling can be seen as a 1st step of validation. It catches easily incompatibilities that would be hard to catch by using the packaged jar directory.
Or maybe there is a standard Java tool that can take 1 jar and 1 classpath and verify that all external types/methods/fields/APIs... referenced by the given jar can be resolved against a given target/classpath and report if some links are broken?
I would personally find it very useful.

**Reply by laeubi** [reply]
But as far as I understand the whole purpose of "Eclipse-Testing-Framework" is to run the precompiled tests?

**Reply by mickaelistria** [reply]
Ah OK. Indeed, it's the goal. However, what particularly are you calling "The Eclipse-Testing-Framework" ? I'm not sure about the scope. Maybe there are things that are outdated in it; but the coretestapplication or uitestapplication are still useful to "just" run tests. By the way, doesn't PDE actually use those?

**Reply by laeubi** [reply]
See this one https://wiki.eclipse.org/Platform-releng/Eclipse_Test_Framework

As far as i understand there are only instructions to run this stuff on the commandline!

**Reply by mickaelistria** [reply]
I think we do need a way to offer running a test from the CLI anyway; and from CLI, it kinds of document as well how to run it with a launch config in the IDE, or with Ant, or with any other scripting tool.
What more do you expect and find missing?

**Reply by laeubi** [reply]
> What more do you expect and find missing?

That's actually the question here... why do we need this "Eclipse Testing Framework", why not just a pom (maybe even profile?) and the mentioned test-mojo with a test runtime as the SDK?

**Reply by mickaelistria** [reply]
> why do we need this "Eclipse Testing Framework", 

We need a way to run existing tests against a given application on different envrionments without rebuilding them every time. This is the purpose of Eclipse Testing Framework. We basically need a way to run tests that does not mandate one build tool or the other (in the end, Eclipse Platform has known several build tools, and the Eclipse Test Framework has still proven useful or at least good enough to never be replaced).

> why not just a pom (maybe even profile?)

Because "just" is not "just" and is probably too much effort for most people to try it when there is actually no problem with the current test framework.
Also because Eclipse-based applications can be built in many ways still (using PDE, using Ant, using Tycho...) and having a solution that any tool can relatively easily consume is valuable.

I'm not saying that we shouldn't move away from this Test Framework; but in the current state of the discussion, I'm not sure what's the problem that's intended to be fixed and what is the better alternative.

> why not just a pom (maybe even profile?) and the mentioned test-mojo with a test runtime as the SDK?

Good quesiton. Please try it and report. If it's better on all aspect, then it will for sure be welcome!

**Reply by akurtakov** [reply]
It is never independent of the build tool, right now it uses Ant. It's not just a pom as back in the days poms (esp in eclipse land) were not that common and everyone understood/wrote Ant and after that no one had nor the will to change it. 
It would be more than welcome change to start using more common way of running tests esp. in the context of the recent fixes to make it work with JUnit 5 suites (https://github.com/eclipse-platform/eclipse.platform.releng/commit/301fa4fc41231abc7358edb7be77161356c785f5)

**Reply by mickaelistria** [reply]
> It is never independent of the build tool, right now it uses Ant.

I personally often use the application from the Eclipse IDE, using an Eclipse Applicaiton launch; and in the past -for JBoss Tools- we've used it with bash...

**Reply by laeubi** [reply]
> I'm not sure what's the problem that's intended to be fixed and what is the better alternative.

The main problem is that once in a while you change something, the verification builds runs fine, and then the SDK build next day gets **KABOOOM** ... then one starts to wonder and luckily  someone step in and say: Have you maybe forgotten to adjust the _(whatever)_ because we do not entirely use the maven build but have our famous (ant?) that works differently. This can include:

- missing to add the test to the suite (mostly unnoticed)
- having missing dependencies somewhere
- ...

then always the next question is, why don't we use maven/tycho to also execute these test, and then its mentioned that "something" will not work. Of course it is the somehow fixed, and forgotten for a while but this is really disturbing ...

So the better alternive would be to not havving a mixture of maven/ant and different configs, but just one technique, maybe even a profile I can activate that switches between "compile+test" and "use sdk+test" ...

**Reply by mickaelistria** [reply]
> missing to add the test to the suite

This is not a particularity of the Test Framework. Similar issues would happen with Tycho/surefire and are more related to maintaining a JUnit test suite.

> having missing dependencies somewhere

Do you have an example of a case where a missing dependency for te test framework wouldn't be equally missing to Tycho/surefire?

> why don't we use maven/tycho to also execute these test

Again, improvements towards that would be welcome. But the improvements and Tycho support need to happen first; then we can start considering deprecating parts of the Test Framework.

To be clear, some requirement for official tests is that they need to run 1. without being rebuilt and 2. with same scripts on all OSs and @sravanlakkimsetti and @SDawley can probably tell more.
I am confident this can be made to work with Tycho on the current state because I managed to implement it for JBoss Tools some 6 or 7 years ago already; however it takes maybe more investigation for Platform to make sure we're not missing important requirement.

**Reply by sravanlakkimsetti** [reply]
As far as I remember eclipse platform started using maven from 4.0. The current build scripts are from 3.x time. The most popular build tool was ant at that time. If the testing can be moved to maven and tycho, yes you can move.

there are couple of requirements
1. Need to run tests in a clean environment. Download the zip file, extract it, install individual test bundle and its dependencies(these are specific to each of the test bundles. for example, pde has very specific dependencies to be installed). Run the tests. Restore the the eclipse instance to same status new download.
2. Need to have timeout for individual test and overall test suite
3. Need to take screenshot of the ui at the time of timeout.(screenshots are listed in the console logs page with extension png)
4. The scripts used are developed quite a long time ago. You can move to tycho if it is simpler that way
5. Need to publish the test results as html so that it can be read easily.

I hope most of these can be achieved by moving to tycho. 



> The main problem is that once in a while you change something, the verification builds runs fine, and then the SDK build next day gets **KABOOOM** ... then one starts to wonder and luckily someone step in and say: Have you maybe forgotten to adjust the _(whatever)_ because we do not entirely use the maven build but have our famous (ant?) that works differently.

This happens other way as well. Sometimes ant will be able to run tests without any issues. but tycho/surefire fails. Since we are using tycho in our verification builds this may not occur that frequently.

**Reply by laeubi** [reply]
@sravanlakkimsetti thanks for this summary, this will help in the investigation!

> This happens other way as well. Sometimes ant will be able to run tests without any issues. but tycho/surefire fails. Since we are using tycho in our verification builds this may not occur that frequently.

Yes, I don't wanted to indicate that test never fail with Tycho because of configuration problems, but it is a bit odd if the only fail for one run but not the other.

**Reply by laeubi** [reply]
@sravanlakkimsetti @mickaelistria @akurtakov  I have one more questions, tests seem to need having a test.xml (see for example https://github.com/eclipse-equinox/equinox/blob/master/bundles/org.eclipse.equinox.http.servlet.tests/test.xml) is this some "magic file" the Eclipse Test Framework scans for? Is this file autogenerated or copied or is it customized for each project (or even on a test basis?)

**Reply by mickaelistria** [reply]
> is this some "magic file" the Eclipse Test Framework scans for? 

Yes. But it's not so magic; it basically defines whether tests are core or UI and what are the test suites to run. It's equivalent to configuration of the tycho-surefire-plugin.

> Is this file autogenerated or copied or is it customized for each project (or even on a test basis?)

It is configured/customized for each project to define at least the test suite to run.

**Comment by akurtakov** [comment]
From releng perspective having POC that shows using few test bundles fetched from the p2 repo being run by maven in clean environment would be good first step for migration.

**Reply by mickaelistria** [reply]
There is https://tycho.eclipseprojects.io/doc/latest/tycho-surefire-plugin/test-mojo.html#testRuntime
One can create a .target that contains a product, create a Tycho module that declare the tests as dependencies and use this testRuntime configuration of tycho-surefire-plugin to see the configured tests (from the .target) running against the product (from the .target as well).
